smarts/sstudio/genscenario.py

```diff
@@ -22,30 +22,125 @@
 
 import collections
+import inspect
 import itertools
 import logging
 import os
 import pickle
-import subprocess
-import sys
+import sqlite3
 from dataclasses import replace
 from pathlib import Path
-from typing import Any, Optional, Sequence, Tuple, Union
+from typing import Any, Dict, List, Optional, Sequence, Tuple, Union
 
 import cloudpickle
+from smarts.core.default_map_builder import find_mapfile_in_dir
+
+from smarts.core.utils.file import file_md5_hash, path2hash, pickle_hash
+from smarts.core.utils.logging import timeit
 
 from . import types
 from .generators import TrafficGenerator
 
 logging.basicConfig(level=logging.WARNING)
 logger = logging.getLogger(__file__)
+logger.setLevel(logging.WARNING)
+
+
+def _check_if_called_externally():
+    frame_info = inspect.stack()[2]
+    module = inspect.getmodule(frame_info[0])
+    if module.__name__ != "smarts.sstudio.genscenario":
+        logger.warning(
+            "",
+            exc_info=DeprecationWarning(
+                "Calling gen_* methods directly is now deprecated. All scenarios must call gen_scenario()."
+            ),
+        )
+
+
+def _build_graph(scenario: types.Scenario, base_dir: str) -> Dict[str, Any]:
+    graph = collections.defaultdict(list)
+
+    graph["map"] = [os.path.join(base_dir, "map", "map.glb")]
+
+    if scenario.map_spec:
+        graph["map_spec"] = [os.path.join(base_dir, "map", "map_spec.pkl")]
+
+    if scenario.traffic:
+        for name, traffic in scenario.traffic.items():
+            ext = "smarts" if traffic.engine == "SMARTS" else "rou"
+            artifact_path = os.path.join(base_dir, "traffic", f"{name}.{ext}.xml")
+            graph["traffic"].append(artifact_path)
+
+    if scenario.ego_missions:
+        graph["ego_missions"] = [os.path.join(base_dir, "missions.pkl")]
+
+    if scenario.social_agent_missions:
+        for name in scenario.social_agent_missions.keys():
+            artifact_path = os.path.join(base_dir, "social_agents", f"{name}.pkl")
+            graph["social_agent_missions"].append(artifact_path)
+
+    if scenario.bubbles:
+        graph["bubbles"] = [os.path.join(base_dir, "bubbles.pkl")]
+
+    if scenario.friction_maps:
+        graph["friction_maps"] = [os.path.join(base_dir, "friction_map.pkl")]
+
+    if scenario.traffic_histories:
+        for dataset in scenario.traffic_histories:
+            assert isinstance(dataset, types.TrafficHistoryDataset)
+            artifact_path = os.path.join(base_dir, f"{dataset.name}.shf")
+            graph["traffic_histories"].append(artifact_path)
+
+    return graph
+
+
+def _needs_build(
+    db_conn: sqlite3.Connection,
+    scenario_obj: Any,
+    artifact_paths: List[str],
+    obj_hash: str,
+    map_needs_build: Optional[bool] = None,
+) -> bool:
+    if scenario_obj is None:
+        return False  # There's no object in the DSL, so nothing to build
+    if map_needs_build:
+        return True  # Passed as an override for artifacts that depend on the map
```

Comment:
You're right, I will pull it out into the conditionals where it is used.

Reply:
On second thought, I think this actually is still useful, as it simplifies what the outer conditional would be. Without it, it would be like

```python
if (scenario.traffic and map_needs_rebuild) or _needs_build(
        db_conn, scenario.traffic, artifact_paths, obj_hash
    ):
    ...
```

rather than

```python
if _needs_build(
        db_conn, scenario.traffic, artifact_paths, obj_hash, map_needs_rebuild
    ):
    ...
```

I'll keep it like it was. Let me know if you disagree or if I'm missing a better way of doing this.

Reply:
Alright, that is a reasonable counter.